### PR TITLE
feat(blocksync): `share.ReadEDS`

### LINKED ISSUE #1104: share: `ReadEDS`
## Context

As per https://github.com/celestiaorg/celestia-node/issues/1099, we want to store EDSes as a whole. For this, the [CAR format](https://ipld.io/specs/transport/car) is chosen.

## Implementation Details

To read an EDS from a stream/file, `ReadEDS` is introduced. Internally, it

- Imports EDS with an empty pre-allocated slice. _NOTE: Size can be taken from DataRoot Row/Col len_
- Wraps given io.Reader with [`BlockReader`](https://github.com/ipld/go-car/blob/master/v2/block_reader.go#L16)
- Reads out blocks one by one and fills up the EDS quadrant via `EDS.SetCell`
  - In total, there should be shares-in-quadrant amount of reads.
- Recomputes and validates via `EDS.Repair`

```go
// ReadEDS reads an EDS quadrant(1/4) from an io.Reader CAR file.
// It expects strictly first EDS quadrant(top left).
// The returned EDS is guaranteed to be full and valid against the DataRoot, otherwise ReadEDS errors.
func ReadEDS(context.Context, io.Reader, DataRoot) (*rsmt2d.ExtendedDataSquare, error)
```

### PR DESCRIPTION
**Pull request by distractedm1nd**
Closes #1104 
Includes implementation and tests

Based on #1139

The base of this PR will change along with 1139 when the feature branch is set up and ready

### CODE REVIEW

File: service/share/eds.go [hunk 1]
```diff
@@ -196,3 +197,40 @@ func rootsToCids(eds *rsmt2d.ExtendedDataSquare) ([]cid.Cid, error) {
...
+
+	odsWidth := len(root.RowsRoots) / 2
+	odsSquareSize := odsWidth * odsWidth
+	shares := make([][]byte, odsSquareSize)
+	// the first quadrant is stored directly after the header,
+	// so we can just read the first odsSquareSize blocks
+	for i := 0; i < odsSquareSize; i++ {
+		block, err := carReader.Next()
+		if err != nil {
+			return nil, fmt.Errorf("failure to read next car entry: %w", err)
```
Comment (vgonkivs) [inline service/share/eds.go:218]:
```suggestion
			return nil, fmt.Errorf("share: failure to read next car entry: %w", err)
```

File: service/share/eds.go [hunk 2]
```diff
@@ -196,3 +197,40 @@ func rootsToCids(eds *rsmt2d.ExtendedDataSquare) ([]cid.Cid, error) {
...
 	return rootCids, nil
 }
+
+// ReadEDS reads the first EDS quadrant (1/4) from an io.Reader CAR file.
+// Only the first quadrant will be read, which represents the original data.
+// The returned EDS is guaranteed to be full and valid against the DataRoot, otherwise ReadEDS errors.
+func ReadEDS(ctx context.Context, r io.Reader, root Root) (*rsmt2d.ExtendedDataSquare, error) {
+	carReader, err := car.NewCarReader(r)
+	if err != nil {
+		return nil, fmt.Errorf("failure to read car file: %w", err)
```
Comment (vgonkivs) [inline service/share/eds.go:207]:
```suggestion
                return nil, fmt.Errorf("failure to read car file: %w", err)
```

File: service/share/eds.go [hunk 3]
```diff
@@ -196,3 +197,40 @@ func rootsToCids(eds *rsmt2d.ExtendedDataSquare) ([]cid.Cid, error) {
...
+
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(odsWidth))
+	eds, err := rsmt2d.ComputeExtendedDataSquare(shares, ipld.DefaultRSMT2DCodec(), tree.Constructor)
+	if err != nil {
+		return nil, fmt.Errorf("failure to compute eds: %w", err)
+	}
+
+	newDah := da.NewDataAvailabilityHeader(eds)
+	if !bytes.Equal(newDah.Hash(), root.Hash()) {
+		return nil, fmt.Errorf("content integrity mismatch: imported root doesn't match expected root")
```
Comment (vgonkivs) [inline service/share/eds.go:233]:
```suggestion
                return nil, fmt.Errorf("share: content integrity mismatch: imported root doesn't match expected root")
```

File: share/eds/eds.go [hunk 4]
```diff
@@ -214,6 +216,43 @@ func rootsToCids(eds *rsmt2d.ExtendedDataSquare) ([]cid.Cid, error) {
...
+	shares := make([][]byte, odsSquareSize)
+	// the first quadrant is stored directly after the header,
+	// so we can just read the first odsSquareSize blocks
+	for i := 0; i < odsSquareSize; i++ {
+		block, err := carReader.Next()
+		if err != nil {
+			return nil, fmt.Errorf("failure to read next car entry: %w", err)
+		}
+		// the stored first quadrant shares are wrapped with the namespace twice.
+		// we cut it off here, because it is added again while importing to the tree below
```
Comment (renaynay) [inline share/eds/eds.go:239]: this needs a TODO for @walldiss when he completes the redundant namespace removal

File: share/eds/eds.go [hunk 5]
```diff
@@ -214,6 +216,43 @@ func rootsToCids(eds *rsmt2d.ExtendedDataSquare) ([]cid.Cid, error) {
...
+
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(odsWidth))
+	eds, err := rsmt2d.ComputeExtendedDataSquare(shares, share.DefaultRSMT2DCodec(), tree.Constructor)
+	if err != nil {
+		return nil, fmt.Errorf("failure to compute eds: %w", err)
+	}
+
+	newDah := da.NewDataAvailabilityHeader(eds)
+	if !bytes.Equal(newDah.Hash(), root.Hash()) {
+		return nil, fmt.Errorf("content integrity mismatch: imported root doesn't match expected root")
```
Comment (renaynay) [inline share/eds/eds.go]: include computed new root and given root pls

File: share/eds/eds.go [hunk 6]
```diff
@@ -214,6 +216,43 @@ func rootsToCids(eds *rsmt2d.ExtendedDataSquare) ([]cid.Cid, error) {
...
+	}
+	return eds, err
```
Comment (renaynay) [inline share/eds/eds.go:253]: return eds, nil ? unless there's a reason to return err here too, but it would be nil.

File: share/eds/eds_test.go [hunk 7]
```diff
@@ -130,6 +131,28 @@ func TestInnerNodeBatchSize(t *testing.T) {
...
+}
+
+func TestReadEDSContentIntegrityMismatch(t *testing.T) {
+	writeRandomEDS(t)
+	dah := da.NewDataAvailabilityHeader(share.RandEDS(t, 4))
+	f := openWrittenEDS(t)
+	defer f.Close()
+
+	_, err := ReadEDS(context.Background(), f, dah)
+	require.Error(t, err)
```
Comment (renaynay) [inline share/eds/eds_test.go:175]: maybe check that the error is the content integrity err

**Review by renaynay — COMMENTED** [review]
tiny nits.

File: share/eds/eds.go [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+	// the first quadrant is stored directly after the header,
+	// so we can just read the first odsSquareSize blocks
+	for i := 0; i < odsSquareSize; i++ {
+		block, err := carReader.Next()
+		if err != nil {
+			return nil, fmt.Errorf("share: failure to read next car entry: %w", err)
+		}
+		// the stored first quadrant shares are wrapped with the namespace twice.
+		// we cut it off here, because it is added again while importing to the tree below
+		shares[i] = block.RawData()[ipld.NamespaceSize:]
```
Comment (Wondertan) [inline share/eds/eds.go:240]: nit: consistency across which namespace size constant is used

**Review by Wondertan — APPROVED** [review]

**Comment by distractedm1nd** [comment]
TODO:
From: @walldiss 
>Nit: here and in other places there is no need to start error message with "failure..." since it is already an error type and will be handled accordingly. Could be just "get root cids: %w"

From: @renaynay 
Make sure all t.Cleanup(cancel)s are called under the definition of the ctx

File: share/eds/eds.go (hunk 4 above)
Comment (distractedm1nd) [reply share/eds/eds.go:240]: done

File: share/eds/eds.go (hunk 5 above)
Comment (distractedm1nd) [reply share/eds/eds.go]: done

File: share/eds/eds.go (hunk 6 above)
Comment (distractedm1nd) [reply share/eds/eds.go]: done

File: share/eds/eds_test.go (hunk 7 above)
Comment (distractedm1nd) [reply share/eds/eds_test.go]: done

File: share/eds/eds.go (hunk 8 above)
Comment (distractedm1nd) [reply share/eds/eds.go]: done

**Review by Wondertan — APPROVED** [review]

**Review by renaynay — APPROVED** [review]

**Pull request merged**